### PR TITLE
driver: eth: Support for lan8651 T1S ETH (with OA TC6 library)

### PR DESCRIPTION
**Pull request by lmajewski**
This patch set provides support for T1S ethernet device - LAN8651.

For communication the approach described in Open Alliance TC6 specification is used. This code (from oa_tc6.[ch]) can be reused with other ICs following this standard.

Test HW:
- nucleo_g491re development board from STM
- Microe LAN8651 module
- LAN8670-USB dongle with debian12 as host

Build:
```
west build -p always -b nucleo_g431rb samples/net/sockets/echo_server/ \
	-- -DCONF_FILE=boards/nucleo_g431rb.conf
```
Test on host PC (one needs to first build echo-client program from net-tools):
```
while :; do ./echo-client -r -i enx001ec0d1cc4d  192.168.3.3; sleep 1; \
	ping -c3 192.168.3.3; done
```

Code required for testing and debugging:
https://github.com/lmajewski/zephyr/commits/t1s-lan8651-nucleo_g431rb-test

### CODE REVIEW

**Comment by lmajewski** [comment]
List of interested developers: @GeorgeCGV @uwearzt @lunn

**Comment by lmajewski** [comment]
Setup LAN8670-USB dongle (to be connected to T1S host) - debian12 (kernel 6.1):
###################################################################
```
cd lan867x-linux-driver-1v0/linux-v6.1.21-support
make
sudo --preserve-env=KBUILD_SIGN_PIN /usr/src/linux-kbuild-6.1/scripts/sign-file sha256 /var/lib/shim-signed/mok/MOK.priv /var/lib/shim-signed/mok/MOK.der microchip_t1s.ko
sudo insmod microchip_t1s.ko enable=1 node_id=0 node_count=8 max_bc=0 burst_timer=128 to_timer=32

Note:
The smsc95xx module shall be loaded first (before the usb dongle is inserted) - otherwise it will
use default PHY driver (not the one for microchip).

To fix it use:
$ echo <usb_device_name> | sudo tee /sys/bus/usb/drivers/smsc95xx/unbind > /dev/null
$ sudo insmod microchip_t1s.ko
$ echo <usb_device_name> | sudo tee /sys/bus/usb/drivers/smsc95xx/bind > /dev/null
```

Design decisions:
#################

- The OA_TC6 routines are using the struct tc6 and struct net_pkt as input.
  It was on purpose to tie it with net_pkt as this is what we do receive to send
  and what is read to upper layers.

- The driver supports half-duplex in terms of the data transmission to/from LAN8651
  via SPI (i.e. when TX is performed, no valid data is read). 

  It can be added, but would increase the design of driver and memory footprint.

  I need to assess if this is worth the effort - for now the driver is as simple
  as possible.

- The driver design differs from Linux one [1], as I had to take into account the
  constrained RAM evironment in Cortex-M4 based CPUs (STM32G491 from nucleo_g491re
  has only 32KiB of RAM to fit network stack and LAN865x driver).

- The PHY part of LAN8651 is very limited - for example the link status is always 1,
  hence there was no good reason to add separate PHY driver (as it was done for
  adin2111).


Test scenarios:
###############

- ARP and ICMP from HOST (ping)
  ping 192.168.3.3

- echo_client (on my Linux HOST PC) -> UDP only
 ` while :; do ./echo-client -r -i enx001ec0d1cc4d  192.168.3.3; sleep 1; ping -c3 192.168.3.3; done`
  
- echo_server run on nucleo_g491:

  `west build -p always -b nucleo_g431rb samples/net/sockets/echo_server/ -- -DCONF_FILE=boards/nucleo_g431rb.conf`

- zperf (CONFIG_NET_ZPERF) cannot be used as it turns out that RAM size is exceeded
  (nucleo_g431re has STM32G491 with only 32 KiB or RAM)


What is missing:
################
- Setup of PLCA parameters from user programs (only DTS is used now)
  (if there would be need a to add the special network management callback 

- Callback to adjust MAC address

- SW based reset - at least on LAN8651 rev B0 simple OA_RESET bit 0
  set has some issues. GPIO based reset works correctly.

Dependencies:
#############
https://github.com/zephyrproject-rtos/zephyr/pull/63239


Open Questions:
#############
- where to place the `oa_tc.[ch]` agnostic files?

File: drivers/ethernet/Kconfig.lan865x [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+config ETH_LAN865X_INIT_PRIORITY
+	int "LAN865X driver init priority"
+	default 72
+	help
+	  LAN865X device driver initialization priority.
+	  Must be initialized after SPI.
+
+
+config ETH_LAN865X_IRQ_THREAD_STACK_SIZE
+	int "Stack size for a thread that processes ADIN IRQ"
```
Comment (jukkar) [inline drivers/ethernet/Kconfig.lan865x:29]: ADIN, is this a typo?

File: drivers/ethernet/eth_lan865x.c [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+	net_if_set_link_addr(iface, ctx->mac_address, sizeof(ctx->mac_address),
+			     NET_LINK_ETHERNET);
+
+	if (ctx->iface == NULL) {
+		ctx->iface = iface;
+	}
+
+	ethernet_init(iface);
+
+	net_eth_carrier_on(iface);
```
Comment (jukkar) [inline drivers/ethernet/eth_lan865x.c:50]: So there is no cable detection?

File: drivers/ethernet/eth_lan865x.c [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+		.mac_address = DT_INST_PROP(inst, local_mac_address),                              \
+		.tx_rx_sem =                                                                       \
+			Z_SEM_INITIALIZER((lan865x_data_##inst).tx_rx_sem, 1, UINT_MAX),   \
+		.int_sem = Z_SEM_INITIALIZER((lan865x_data_##inst).int_sem, 0, UINT_MAX),  \
+		.tc6 = &oa_tc6_##inst \
+	};                                                                                         \
+                                                                                                   \
+	ETH_NET_DEVICE_DT_INST_DEFINE(inst, lan865x_init, NULL, &lan865x_data_##inst, \
+				      &lan865x_config_##inst, CONFIG_ETH_INIT_PRIORITY,       \
+				      &lan865x_api_func, NET_ETH_MTU);
```
Comment (jukkar) [inline drivers/ethernet/eth_lan865x.c:549]: Minor nit but could you adjust the \\ at the end of the line to be in the same line, it looks hurts to look at it.

File: drivers/ethernet/oa_tc6.c [hunk 4]
```diff
@@ -0,0 +1,356 @@
+/*
```
Comment (jukkar) [inline drivers/ethernet/oa_tc6.c:1]: Please place the oa_tc6.c into a separate commit (could be introduced before the driver commit).

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# Generic networking options
```
Comment (jukkar) [inline samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:1]: This sample board config should be in a separate commit.

Please only place here board specific stuff so that the main prj.conf in that sample will be the master information and here we just tweak things needed for this board. Do not place config options specific to your local setup here (like IP addresses etc).

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+CONFIG_NET_CONFIG_SETTINGS=y
+CONFIG_NET_CONFIG_NEED_IPV4=y
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.168.3.3"
+CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.168.3.1"
+
+# How many client can connect to echo-server simultaneously
+CONFIG_NET_SAMPLE_NUM_HANDLERS=1
+
+# LAN865x T1S driver specific
+CONFIG_NET_L2_ETHERNET=y
```
Comment (jukkar) [inline samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:49]: This should be set in the board defaults and not here.

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+CONFIG_NET_PROMISCUOUS_MODE=y
```
Comment (jukkar) [inline samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:50]: Why would we need to enable promiscuous mode by default?

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.overlay [hunk 8]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (jukkar) [inline samples/net/sockets/echo_server/boards/nucleo_g431rb.overlay:1]: This can be in the same commit as the sample board config.

**Review by jukkar — COMMENTED** [review]
Looks quite resonable. Please fix the compliance issues, there are several whitespace issues in the code.
Please change the driver commit subject to
`driver: eth: Support for lan8651 T1S ETH`

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf (hunk 6 above)
Comment (lmajewski) [reply samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:49]: This board don't have ethernet connector. This is required only when MicroE's add-in module is used.

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf (hunk 7 above)
Comment (lmajewski) [reply samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:50]: That was a leftover from testing - I will comment this out (to show that LAN865x supports it)

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.overlay (hunk 8 above)
Comment (lmajewski) [reply samples/net/sockets/echo_server/boards/nucleo_g431rb.overlay:1]: Ok, will make those two files as a separate commit.

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf (hunk 5 above)
Comment (lmajewski) [reply samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:1]: The `prj.conf` file needs changes as shown in `nucleo_g431rb.conf`, otherwise the sample will not build (excess of on-board RAM).

IMHO, it would be better to have a complete setup there for this particular development board, so other users can reproduce this setup easily.

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf (hunk 6 above)
Comment (jukkar) [reply samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:49]: > This board don't have ethernet connector. This is required only when MicroE's add-in module is used.

Then should we have this patch at all as the other users of this board but without this add-in module would get incorrect configuration.

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf (hunk 5 above)
Comment (jukkar) [reply samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:1]: Sorry, but this is not a way here. The prj.conf will be the master of the config and the config in board/ directory can then add or adjust the configuration parameters. This way if any new config options are added in prj.conf, the compilation for this board will automatically use them and we do not need to adjust two files.

**Review by jukkar — CHANGES_REQUESTED** [review]
Please check the compliance issues

File: samples/net/sockets/echo_server/boards/nucleo_g431rb.conf (hunk 6 above)
Comment (lmajewski) [reply samples/net/sockets/echo_server/boards/nucleo_g431rb.conf:49]: There are "shields" defined for many other boards - like `MicroE ETH click` - `zephyr/boards/shields/mikroe_eth_click`

I could follow that path - the `nucleo_g431re` supports `arduino header`.

Although, in my case I only do have "wire" connected pins between `nucleo`  and `Micro BUS LAN8651` module, I can pretend that I do have a special arduino "shield" for this chip.

Comment (erwango) [reply samples/net/sockets/echo_server/boards/nucleo_g431rb.conf]: > I can pretend that I do have a special arduino "shield" for this chip.

No, you can't. We only maintain actual shields.

File: boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay [hunk 9]
```diff
@@ -0,0 +1,23 @@
+&arduino_spi {
```
Comment (erwango) [inline boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay:1]: Nack on this. If there's no arduino connector, don't make up one

**Review by erwango — DISMISSED** [review]

File: boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay (hunk 9 above)
Comment (lmajewski) [reply boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay:1]: @erwango - I've added the MicroE's shield in the newest version of this patch set.
I've also followed the approach from `zephyr/boards/shields/mikroe_wifi_bt_click` shield, which allows Arduino operation as well.

My goal is to provide and upstream the setup, so either me or other peers can easily test the T1S solution.

Comment (erwango) [reply boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay:1]: > shield, which allows Arduino operation as well.

Ok, then we should review this as well. It won't be possible to maintain and provide support on each and every fantasy connection that one can imagine on his/her desk.

Comment (lmajewski) [reply boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay:1]: @erwango - I would like to also bring up the issue: the need to document the way one would be able to reproduce the setup. Of course - as I did with this PR - I can add detailed description here, but after this code is pulled - the info is gone.

I would prefer to keep this information in some form. Do you have any idea of how this can be done?

Comment (erwango) [reply boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay:1]: W/o better option, I'd propose 2 options:
- Move this configuration in a board.overlay file in the sample
- Document set up used for test in commit message

Comment (lmajewski) [reply boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay:1]: I think that, I could also provide link to my github account with those configuration files.

Adding `board.overlay` to `echo_server` sample is tricky, as it will not work without the "shield" and can cause regression on the board itself (this board would not be used in this sample as it doesn't have ETH on it).

Comment (lmajewski) [reply boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay:1]: @erwango @jukkar - then I will just add the OA_TC6 code and eth_lan865x.

For extra *.overlay and *.conf setup I will add link to my github account and provide detailed description in this PR.

**Comment by GeorgeCGV** [comment]
The OA could be interesting for ADIN MAC ICs as they support ADI SPI and OA. One of the main benefits is the full-duplex support when using OA at a lower clock.

Maybe ADI won't be against porting their OA implementation. It implements full-duplex. Probably, @amiclaus @MaureenHelm can clarify that.

**Comment by lmajewski** [comment]
> The OA could be interesting for ADIN MAC ICs as they support ADI SPI and OA. One of the main benefits is the full-duplex support when using OA at a lower clock.
> 
> Maybe ADI won't be against porting their OA implementation. It implements full-duplex. Probably, @amiclaus @MaureenHelm can clarify that.

@GeorgeCGV  As fair as I can tell - ADI's SoC (ADIN2111) only supports their (i.e. non-OA) communication.

The full-duplex transmission of OA TC6 via SPI is also a valid option, yes - I'm also thinking  how to implement it.

**Comment by GeorgeCGV** [comment]
> @GeorgeCGV As fair as I can tell - ADI's SoC (ADIN2111) only supports their (i.e. non-OA) communication.

Based on [product-overview](https://www.analog.com/en/products/adin2111.html#product-overview):

> The 4-wire SPI for communication with the host can be configured to OPEN Alliance SPI or generic SPI. Both modes support optional data protection or cyclic redundancy check (CRC).

Also, implementation is present within [ADIN2111 Non Operating System Driver](https://www.analog.com/en/products/adin2111.html#product-requirement).

**Comment by lmajewski** [comment]
> > @GeorgeCGV As fair as I can tell - ADI's SoC (ADIN2111) only supports their (i.e. non-OA) communication.
> 
> Based on [product-overview](https://www.analog.com/en/products/adin2111.html#product-overview):
> 
> > The 4-wire SPI for communication with the host can be configured to OPEN Alliance SPI or generic SPI. Both modes support optional data protection or cyclic redundancy check (CRC).

Sorry, my mistake I have been imprecise - I meant that the driver in Zephyr now only supports generic SPI.

> 
> Also, implementation is present within [ADIN2111 Non Operating System Driver](https://www.analog.com/en/products/adin2111.html#product-requirement).

**Comment by GeorgeCGV** [comment]
@lmajewski yes, the driver supports generic SPI only

**Review by jukkar — DISMISSED** [review]

File: boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay (hunk 9 above)
Comment (lmajewski) [reply boards/shields/mikroe_twowire_eth_click/mikroe_twowire_eth_click_arduino.overlay]: @erwango - I've removed the shield code. Could you re-review the code (as I don't have the proper re-review mark next to yours name as I had for @jukkar)?

**Comment by carlescufi** [comment]
@erwango can you take another look?

**Comment by erwango** [comment]
Sorry, but I shouldn't be the assignee on such PR no shield, noSTM32). Assigning @rlubos as PR has already been reviewed by @jukkar.

**Comment by lmajewski** [comment]
> Sorry, but I shouldn't be the assignee on such PR no shield, noSTM32). Assigning @rlubos as PR has already been reviewed by @jukkar.

The aforementioned "shield" has been removed with this PR, as it has been pointed out that Zephyr will not accept non compatible combination of base board (nucleo) and MicroBus shield.

Moreover, I would appreciate if we could finish with review of this PR.

**Comment by jukkar** [comment]
@erwango if you are not reviewing this any more, please remove your -1

**Comment by lmajewski** [comment]
@jukkar - I had to rebase those patches on top of newest origin/main branch. The only change was with Kconfig.

**Comment by lmajewski** [comment]
@jukkar - Please find list of enhancements for this driver (as I had to rebase it on top of newest 'main' branch):

- Add support for setting MAC address
- Reduce IRQ stack size
- Check if received frame shall be dropped (based on footer information)
- Optimize 'sbo' and 'ebo' usage when receiving packets
- Handle special case when frame ends and starts in the same chunk

File: drivers/ethernet/eth_lan865x.c [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+}
+
+static int lan865x_check_spi(const struct device *dev)
+{
+	struct lan865x_data *ctx = dev->data;
+	uint32_t val;
+	int ret;
+
+	ret = oa_tc6_reg_read(ctx->tc6, LAN865x_DEVID, &val);
+	if (ret < 0)
```
Comment (jukkar) [inline drivers/ethernet/eth_lan865x.c:142]: Missing {}

Comment (jukkar) [reply drivers/ethernet/eth_lan865x.c:142]: No worries, the rest of the PR LGTM.

**Review by jukkar — DISMISSED** [review]

File: drivers/ethernet/eth_lan865x.c [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+
+	return 0;
+}
+/* Implementation of pseudo code from AN1760 - END */
+
+static int lan865x_config_plca(const struct device *dev, uint8_t node_id,
+			       uint8_t node_cnt, uint8_t burst_cnt, uint8_t burst_timer)
+{
+	struct lan865x_data *ctx = dev->data;
+	uint32_t regVal;
```
Comment (jfischer-no) [inline drivers/ethernet/eth_lan865x.c:242]: regVal name looks suspicious, is it to follow specific name from standard? Otherwise it should not be camelCase.

**Review by jfischer-no — COMMENTED** [review]
Nice :+1:

File: drivers/ethernet/eth_lan865x.c (hunk 11 above)
Comment (lmajewski) [reply drivers/ethernet/eth_lan865x.c:242]: Ok, I will fix it

**Review by jfischer-no — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
